dhall/src/Dhall/TypeCheck.hs

```diff
@@ -387,20 +387,24 @@ typeWithA tpa = loop
                             Left (TypeError ctx x err) )
                 return (App List t)
             _ -> Left (TypeError ctx e MissingListType)
-    loop ctx e@(ListLit (Just t ) xs) = do
-        s <- fmap Dhall.Core.normalize (loop ctx t)
-        case s of
-            Const Type -> return ()
-            _ -> Left (TypeError ctx e (InvalidListType t))
+    loop ctx e@(ListLit (Just t0 ) xs) = do
+        let nf_t0 = Dhall.Core.normalize t0
+        t1 <- case nf_t0 of
+            App List t1 -> do
+                s <- fmap Dhall.Core.normalize (loop ctx t1)
+                case s of
+                    Const Type -> return t1
+                    _ -> Left (TypeError ctx e (InvalidListType t1))
+            _ -> Left (TypeError ctx e (InvalidListType t0)) -- TODO
```

Comment:
Should I make a new parser error for this case or is there something I could re-use?

Reply:
I think this line and the line above it should both use the `InvalidListType` error type.  However, the explanation for the `InvalidListType` should now explain the error in terms of the entire type annotation for the list being wrong instead of just the element type being wrong